### PR TITLE
Tokenize language messages and simplify loading

### PR DESCRIPTION
**Pull request by kcgen**
### French, German, Italian, Polish, Russian, Spanish users

You can test this as follows:

1. Download the package for your OS:
    - Linux - https://github.com/dosbox-staging/dosbox-staging/suites/3318325312/artifacts/77718770
    - Windows - https://github.com/dosbox-staging/dosbox-staging/suites/3318325313/artifacts/77719484
    - macOS - https://github.com/dosbox-staging/dosbox-staging/suites/3318325311/artifacts/77719009

2. Extract the package

3. Confirm your existing language/translation approach works as-is:
    - Using the `-lang /path/to/file.lng` or the `language = path/to/file.lng` conf setting.
    - Run `INTRO` at the Z:\> prompt, to see your existing translated messages.
    - Confirm that your keyboard modifiers are correct: `Ctrl` and `Alt` on PC, `Cmd` and `Opt` on macOS.

4. Confirm this PR's improved language approach also works:
    - Copy the package's `translations` directory (macOS users: the .dmg has it under `Resources/translations`) into your DOSBox Staging `conf` directory.
    - Your `conf` directory is logged to the console in the first `CONFIG:` line. For example:
    ``` text
    dosbox-staging version 0.78.0-alpha-92-g3960d0695
    ---
    CONFIG: Loading primary file /home/user/.config/dosbox/...
    DISPLAY: Initialized 1280x960 window-mode on 3840x2160 display-0
    ```

5. Launch with: `dosbox -lang de` (or `es`, `it`, `pl`, `ru`).  
6. Also try a similar conf setting:
    ``` ini
    [dosbox]
    language = de
    ```

(end of test procedure)

----

As @dreamer noted [here](https://github.com/dosbox-staging/dosbox-staging/issues/884#issuecomment-877085992), commit 4e70670 broke the translation system.

This PR fixes it by replacing the inline strings with formatting tokens, which are now formatted during printing. For example:

Because this PR re-generates the language files, it turns out to be quite large - however all of the `.lng` files are generated, and therefore can be ignored and the `.txt` equivalents can be reviewed instead.

This PR also include the `translations/` directory (containing non-dialect `.lng` files) in the CI-generated release packages, to ensure users have the latest `.lng` files in tandem with the binary.  

Users can copy this `translations` directory into their config path for easier language loading. This PR also notes this with usage examples in the README.

For example, on launch:
``` shell
dosbox -lang de
dosbox -lang es
dosbox -lang fr
dosbox -lang it
dosbox -lang pl
dosbox -lang ru
```

Or in the conf file:
```ini
[dosbox]
language = de
# es, fr, it, pl, ru
```

macOS vs. PC hotkeys are also handled properly.  For example, the same language files can be loaded on a macOS, and the hotkeys listed in the `INTRO` screens will show as `Opt` and `Cmd`.

### CODE REVIEW

**Comment by kcgen** [comment]
To find the instances that need to be formatted during printing, we can use:

`grep -B7 %s+ contrib/translations/utf-8/en/en-0.77.0-alpha.txt | grep '%\|^:'`

Which produces:
``` text
:PROGRAM_INTRO_SPECIAL
%s+Enter  Switch between fullscreen and window mode.
%s+Pause  Pause/Unpause emulator.
%s+F1   %s Start the keymapper.
%s+F4   %s Swap mounted disk image, update directory cache for all drives.
%s+F5   %s Save a screenshot.
%s+F6   %s Start/Stop recording sound output to a wave file.
%s+F7   %s Start/Stop recording video output to a zmbv file.
%s+F9   %s Shutdown emulator.
%s+F10  %s Capture/Release the mouse.
%s+F11  %s Slow down emulation.
%s+F12  %s Speed up emulation.
%s+F12    Unlock speed (turbo button/fast forward).

:PROGRAM_BOOT_PRINT_ERROR
by pressing %s+F4, and -l specifies the mounted drive to boot from.  If
  - %s+F4 swaps & mounts the next CDROM-SET or IMAGEFILE, if provided.

:SHELL_STARTUP_BEGIN
║ Welcome to DOSBox Staging %-40s ║
║ To adjust the emulated CPU speed, use %s+F11 and %s+F12.%s%s       ║
║ To activate the keymapper %s+F1.%s                                 ║

:SHELL_STARTUP_CGA
║ (%s+)F11 changes hue; %s+%s+F11 selects early/late CGA model.%s ║

:SHELL_STARTUP_CGA_MONO
║ and %s+F11 to change contrast/brightness settings.                ║

:SHELL_STARTUP_DEBUG
║ Press %s+Pause to enter the debugger or start the exe with DEBUG. ║
```
---

Each use of these labels can then be grepped out of sources, paying attention when we "MSG_Get(..)". 

`git grep PROGRAM_INTRO_SPECIAL *.cpp`

``` text
src/dos/dos_programs.cpp:               WriteOut(MSG_Get("PROGRAM_INTRO_SPECIAL"), MMOD2_NAME,
(other instances are MSG_Add or in sdl_gui, which is unused code).
```

`git grep PROGRAM_BOOT_PRINT_ERROR *.cpp`
``` text
src/dos/dos_programs.cpp:               WriteOut(MSG_Get("PROGRAM_BOOT_PRINT_ERROR"), PRIMARY_MOD_NAME);
```

`git grep PROGRAM_BOOT_PRINT_ERROR *.cpp`
``` text
src/dos/dos_programs.cpp:               WriteOut(MSG_Get("PROGRAM_BOOT_PRINT_ERROR"), PRIMARY_MOD_NAME)
```

`git grep SHELL_STARTUP_BEGIN *.cpp`
``` text
rc/shell/shell.cpp:                    WriteOut(MSG_Get("SHELL_STARTUP_BEGIN"),
```

`git grep SHELL_STARTUP_CGA *.cpp`
``` text
src/shell/shell.cpp:                                    WriteOut(MSG_Get("SHELL_STARTUP_CGA_MONO"),
src/shell/shell.cpp:                                    WriteOut(MSG_Get("SHELL_STARTUP_CGA"),
```

`git grep SHELL_STARTUP_DEBUG *.cpp`
``` text
src/shell/shell.cpp:                    WriteOut(MSG_Get("SHELL_STARTUP_DEBUG"), MMOD2_NAME);
```

**Comment by kcgen** [comment]
Moving out of draft:
 - Tested all languages
 - Tested with macOS "switch" to set true, and confirmed macOS buttons are shown.
 - Tested positive and negative scenarios (missing lang file, untranslated languages, etc.)

**Comment by kcgen** [comment]
**Future additions:**

 - Add a version-control tag to the .lng files when generating, and check these at runtime to ensure a match (and prompt the user that their lng files are outdated).

 - Search for the `translations` directory in the executable's path (first in the priority ordering) to ensure the as-extracted release package uses its corresponding .lng files. (when C++17 is opened up, as it should make discovering the executable's path easier).

**Comment by Burrito78** [comment]
@kcgen Thanks for working on this!! Exactly what i need to finish the german translation properly.

I tested everything as you told in the first post and everything is working except here:

"3. Confirm your existing language/translation approach works as-is:

Using the -lang /path/to/file.lng or the language = conf setting."

The command line "-lang" parameter is working, but setting "language = es" for example still gives me english texts.

This is with the build you linked and a fresh install.

**Comment by kcgen** [comment]
Thank you, @Burrito78 for testing this out. 

> @kcgen Thanks for working on this!! Exactly what i need to finish the german translation properly.

:+1: 

> The command line "-lang" parameter is working

:+1: 

> setting "language = es" for example still gives me english texts. This is with the build you linked and a fresh install. I didn't provide any .lng files myself.

From this failed loading, can you paste the `**LANG:**` output from your console? 

It should look something like this (examples from a Linux system):

``` text
dosbox-staging version 0.78.0-alpha-95-g4bba3ae04
...
LANG: Searching config path for: es.lng
LANG: Language file /home/burrito78/.config/dosbox/translations/es.lng not found, skipping
...


When it succeeds, it will show:

``` text
LANG: Searching config path for: es.lng
LANG: Loaded language file: /home/burrito78/.config/dosbox/translations/es.lng
```

The `.dmg` package contains a `Resources/translations` directory; however I'm not sure how you can find or get this from the .dmg package (hopefully you can access it?).

The dmg's `translations` needs to be dragged into your DOSBox Staging conf path to match the above.



Curious if this helps - and if not, please copy any errors or messages - thank you @Burrito78 !

**Comment by Burrito78** [comment]
@kcgen Thanks for the fast and thorough reply!

From your first post i was under the impression that there were 2 "legacy" ways to open a language file and 2 "new" ways.
So i assumed that for "legacy" to work, nothing had to be copied. 

So if i don't copy anything and set "language = es" i get an error message in the log:
```
LANG: Searching config path for: es.lng
LANG: Language file /Users/Alex/Library/Preferences/DOSBox/translations/es.lng not found, skipping
```

When i copy the translations folder out of the dmg and into the conf directory, everything is working ok!

Thanks for clearing that up!

**Review by Burrito78 — APPROVED** [review]
Tested all the different options and different language files. All working fine. Cmd modifier is shown on macOS as expected and is working.

**Comment by kcgen** [comment]
@Burrito78, the German translation is now included!

You can grab the new .DMG from the macOS link in the message at the top, and its `translations` includes the new **`de.lng`** that you can copy over into your config translations directory. 

Curious how it goes!

PS - the intro banner changes with machine type (in the conf).  Try `machine = ` ... `cga`, `cga_mono`, and `hercules`.

**Comment by kcgen** [comment]
A couple German intros, courtesy of @Burrito78's translation:

**Comment by kcgen** [comment]
@bmunger , @kklobe ,  I think this is ready for your critical eyes and feedback.

Use this filtering link: https://github.com/dosbox-staging/dosbox-staging/pull/1156/files?file-filters%5B%5D=.cpp
 to collapse all files except for the source code.

**Comment by Burrito78** [comment]
Thanks for directly adding the german translation!

Man, still a lot of formatting to improve! Already making a list for the next update!

**Comment by Burrito78** [comment]
> Curious how it goes!
> 
> PS - the intro banner changes with machine type (in the conf). Try `machine = ` ... `cga`, `cga_mono`, and `hercules`.

@kcgen Thanks for directly adding the german translation! Enabling it was no problem. Exciting! Different machine types change the intro banner, yes. Tried it and it worked.

Man, still a lot of formatting to improve! Already making a list for the next update!

**Comment by kcgen** [comment]
> Already making a list for the next update!

Yes! There are also updated english messages; you can see the entire new file:
 - https://github.com/dosbox-staging/dosbox-staging/blob/master/contrib/translations/en/en_US.lng

And more helpful, you can see how it changed versus the file you original translated in this commit: https://github.com/dosbox-staging/dosbox-staging/pull/1156/commits/b499073390e770b125a2c5fc3445cc8dd1c7d17e

**Comment by Burrito78** [comment]
> And more helpful, you can see how it changed versus the file you originally translated in this commit: [b499073](https://github.com/dosbox-staging/dosbox-staging/commit/b499073390e770b125a2c5fc3445cc8dd1c7d17e)

@kcgen Now THAT is REALLY useful. Thanks for posting this link!!

File: src/misc/messages.cpp [hunk 1]
```diff
@@ -16,20 +16,22 @@
...
-using namespace std;
 
+#include "control.h"
+#include "cross.h"
+#include "fs_utils.h"
+#include "string_utils.h"
+#include "setup.h"
+#include "support.h"
 
+using namespace std;
```
Comment (bmunger) [inline src/misc/messages.cpp:34]: Should we really be doing this?

**Review by bmunger — COMMENTED** [review]
I do find the changes to help messages harder to follow in the code after this change.  Not sure what the benefit is overall though.

File: src/misc/messages.cpp (hunk 1 above)
Comment (kcgen) [reply src/misc/messages.cpp:34]: nope :-)

**Comment by kcgen** [comment]
> I do find the changes to help messages harder to follow in the code after this change. Not sure what the benefit is overall though.

Yes, it's doing a lot more work - I didn't explain how the previous code works, but if you run `dosbox -lang ARG`, that ARG needs to be an exact path.  `dosbox -lang /path/to/my/exact/es_ES.lng`.

The changes still support that, but additionally support the much more convenient  `-lang ru` syntax. The changes will look for the `<config_path>/translations/<lang>.lng` file, where the config_path itself is cross-platform. 

I've added a discussion type comment to this function describing both approaches - hopefully that helps!

Thanks for the feedback @bmunger -- "maintain the maintainers" is job number 1 - and that means understanding what's going on.

File: src/misc/messages.cpp (hunk 1 above)
Comment (kcgen) [reply src/misc/messages.cpp]: Fixed.

**Comment by bmunger** [comment]
I should have explained what I meant more specifically.  Replacing the inline strings with the string format specifier makes it easier to miss within the ANSI color codes.  It's more visible in an actual IDE instead of Github, but I understand why you chose to handle it that way.

**Review by bmunger — APPROVED** [review]

**Comment by kcgen** [comment]
> Replacing the inline strings with the string format specifier makes it easier to miss within the ANSI color codes.

Oh, yes - it is.  I think that when we move to gettext, it should solve this (and more)!

Thanks for the review; with all the tests passing, and also passing UBSAN and ASAN, I think we can merge it now.

**Comment by Kappa971** [comment]
I have done all the tests listed and it seems to work (the it.lng file is the old one but I think it is the same to take the test). The problems I encountered are:
1. you still need to specify the code page 850 via the keyb command (keyb it 850) otherwise some characters are not displayed correctly. I am attaching an example screenshot:
Immagine 2021-07-24 124809

Here with keyb it 850:
Immagine 2021-07-24 124853

2. More than a problem, I would say that it is a feature to be added, ie that the configuration file remains in English

**Pull request merged**